### PR TITLE
SCP-3470: `plutus-ledger-api` support for CIPs 31-33

### PR DESCRIPTION
**Pull request by michaelpj**
This adds the additional fields to the V2 `TxInfo` that are needed for CIPs 31-33.

In particular:
- A new field in `TxInfo` for the reference inputs.
- A new field in `TxOut` for the datum/inline-datum/nothing: since this is a tristate thing I wrote a special datatype. That's also what the ledger does.
- A new field in `TxOut` for the reference script.

### CODE REVIEW

**Review by ak3n — APPROVED** [review]

File: plutus-ledger-api/src/Plutus/V1/Ledger/Tx.hs [hunk 1]
```diff
@@ -207,7 +210,7 @@ isPubKeyOut = isJust . txOutPubKey
 
 -- | Whether the output is a pay-to-script output.
 isPayToScriptOut :: TxOut -> Bool
-isPayToScriptOut = isJust . txOutDatum
+isPayToScriptOut = isJust . txOutValidatorHash
```
Comment (bezirg) [inline plutus-ledger-api/src/Plutus/V1/Ledger/Tx.hs:213]: Isn't this change unrelated to the CIP? It looks more like a fix to make the type of the address (p2sh or normal) to be determined not by the existence of the datum, but by the "deserialization" / format of the address

File: plutus-ledger-api/src/Plutus/V2/Ledger/Tx.hs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    (OutputDatumHash dh) == (OutputDatumHash dh') = dh PlutusTx.== dh'
+    (OutputDatum d) == (OutputDatum d')           = d PlutusTx.== d'
+    _ == _                                        = False
+
+instance Pretty OutputDatum where
+    pretty NoOutputDatum        = "no datum"
+    pretty (OutputDatumHash dh) = "datum hash: " <+> pretty dh
+    pretty (OutputDatum d)      = "inline datum : " <+> pretty d
+
+-- | A transaction output, consisting of a target address, a value, and optionally a datum hash.
```
Comment (bezirg) [inline plutus-ledger-api/src/Plutus/V2/Ledger/Tx.hs:79]: the haddock comment needs updating

File: plutus-ledger-api/src/Plutus/V2/Ledger/Tx.hs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    pretty TxOut{txOutAddress, txOutValue, txOutDatum, txOutReferenceScript} =
+                hang 2 $ vsep ["-" <+> pretty txOutValue <+> "addressed to", pretty txOutAddress, "with datum", pretty txOutDatum, "with referenceScript", pretty txOutReferenceScript]
+
+instance PlutusTx.Eq TxOut where
+    {-# INLINABLE (==) #-}
+    (TxOut txOutAddress txOutValue txOutDatum txOutRefScript) == (TxOut txOutAddress' txOutValue' txOutDatum' txOutRefScript') =
+        txOutAddress PlutusTx.== txOutAddress'
+        PlutusTx.&& txOutValue PlutusTx.== txOutValue'
+        PlutusTx.&& txOutDatum PlutusTx.== txOutDatum'
+        PlutusTx.&& txOutRefScript PlutusTx.== txOutRefScript'
```
Comment (bezirg) [inline plutus-ledger-api/src/Plutus/V2/Ledger/Tx.hs:99]: is this more strict than necessary? I thought providing a reference script is kind of optional? i thought the hash+datum is the enough to be the determined factor of an (Eq TxOut). I am not sure about this.

**Review by bezirg — COMMENTED** [review]
Looks ok with some minor changes

File: plutus-ledger-api/src/Plutus/V1/Ledger/Tx.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-ledger-api/src/Plutus/V1/Ledger/Tx.hs:213]: Yeah, its a bug fix. The previous version was wrong.

File: plutus-ledger-api/src/Plutus/V2/Ledger/Tx.hs (hunk 3 above)
Comment (michaelpj) [reply plutus-ledger-api/src/Plutus/V2/Ledger/Tx.hs:99]: Hmm, I don't think so. The `TxOut` is what it is: it's optional to have bits of it, but those still form part of its identity if they're there!

**Pull request merged**